# Oxford Nanopore Read Processing.





## 1. Basecalling
During the process of sequencing DNA using nanopores, variations in electrical current caused by the DNA strand passing through the nanopore are captured and stored as raw data in FAST5 files. Basecaller algorithms like ONT’s Guppy ™ software and ONT’s Dorado™ software analyse the raw signal data within these FAST5 files to translate the sequence of bases into FASTQ files. A FASTQ file contains information on the sequence of nucleotides and the quality (confidence) score associated with each nucleotide.
You can perform basecalling using either Dorado or Guppy. Dorado is the preferred basecaller but does not support all the output configurations from ONT platforms. In that case, Guppy still provides a good alternative to perform basecalling. This is really only an issue if you have used kit12 chemistry with R10.4 flow cells. Dorado does support basecalling for kit10 and 9.4.1 flow cells and kit14 and R10.4.1 flow cells or newer.

### Basecalling with Dorado v0.3.1

#### Convert with pod5

We'll use Dorado to convert the raw output from your sequencing run to fastq data. Often you receive both fast5 and fastq files. In such cases basecalling is performed while output was generated. However, it's best to redo the basecalling using a hyper accurate model.

First we need to convert the fast5 files for POD5 using pod5.

The following piece of code retrieves the name of ith fast5 file in your FAST5_DIR and creates a new file extension for the POD5 file. Then pod5 converts the ith FAST5 file in the array to POD5. This will create the same number of pod5 files as fast5 files.

```bash
#!/bin/bash
### IMPORTANT: manually change to number of arrays (-a) to the number of file in your fast5 directory

## Load modules
#module load GCCcore/11.3.0
#module load Python/3.10.4

## Set directories
FAST5_DIR=path/to/input/dir/fast5
OUTPUT_DIR=path/to/output/pod5

##POD5 is required- you can use pip install pod5
pod5 convert fast5 $FAST5_DIR/*.fast5 --output $OUTPUT_DIR --one-to-one $FAST5_DIR
````

#### Basecalling using Dorado
##### Splitting the .pod5 files.
As reccommended by Dorado, we'll subset the .pod5 files into per-channel unique .pod5 files (this also speeds up duplex basecalling; ([see here](https://github.com/nanoporetech/dorado#improving-the-speed-of-duplex-basecalling)). The per-channel pod5 files were then basecalled for duplex and simplex reads *simultaneously* using dorado duplex function, with the model specified as dna_r10.4.1_e8.2_400bps_sup@v4.1.0 (super accurate basecalling model).

An explanation of how the duplex function works can be read [here](https://github.com/nanoporetech/dorado?tab=readme-ov-file#duplex)
```bash
#!/bin/bash
# Load necessary modules
module load GCCcore/11.3.0
module load Python/3.10.4

# Variables
POD5_DIR=/path/to/pod5
OUTPUT_DIR=/path/to/output/dir/pod5_by_channel

SUMMARY_FILE=$POD5_DIR/summary.tsv

# Create output directory if it doesn't exist
mkdir -p $OUTPUT_DIR

# Step 1: Generate a summary file listing read IDs split by channel
# This summary file will include the read ID and channel information
pod5 view $POD5_DIR --include "read_id,channel" --output $SUMMARY_FILE

# Step 2: Split the reads in the POD5 files and subset them by channel
# This is based on the 'summary' file created in Step 1
pod5 subset $POD5_DIR --summary $SUMMARY_FILE --columns channel --output $OUTPUT_DIR

````

##### Basecalling and Generating BAM Files
```bash
#!/bin/bash
# Load GPU modules (adjust to your system)
module load GCC/10.2.0
module load CUDA/11.1.1

# Paths
dorado=/path/to/dorado
MODEL=/path/to/dna_r10.4.1_e8.2_400bps_sup@v4.1.0
POD5_DIR=/path/to/pod5_files
BAM_DIR=/path/to/output/bam_files

# Parallel processing setup
i=${SLURM_ARRAY_TASK_ID}
NAME=$(ls $POD5_DIR/*.pod5 | head -n $i | tail -n 1 | sed -e 's/\.pod5$//' | xargs -n 1 basename)
TMP_DIR=$BAM_DIR/$NAME/

# Create temporary directory and copy file
mkdir -p $TMP_DIR
cp $POD5_DIR/$NAME.pod5 $TMP_DIR

# Basecalling with Dorado
$dorado duplex $MODEL $TMP_DIR > $BAM_DIR/$NAME.bam

# Cleanup
rm -r $TMP_DIR
````
##### Distinguishing between simplex/duplex reads
The dx tag in the BAM record for each read can be used to distinguish between simplex and duplex reads
•	dx:i:1 for duplex reads.
•	dx:i:0 for simplex reads which don't have duplex offsprings.
•	dx:i:-1 for simplex reads which have duplex offsprings.

**Note:** I reccommend merging your *.bam files prior to running this step!

```bash
# Define paths to tools and directories
BAM_DIR=/path/to/merged/bam/dir
OUTPUT_DIR=/path/to/output/dir

# Example 1: Extract duplex reads (dx:i:1)
samtools view ${BAM_DIR}/merged.bam | awk -F '\t' '{for(i=12;i<=NF;i++){if($i ~ /^dx:i:1$/){print; next}}}' | samtools view -b -o ${OUTPUT_DIR}/duplex_output.bam

# Example 2: Extract simplex reads (dx:i:0 and dx:i:-1)
samtools view ${BAM_DIR}/merged.bam | awk -F '\t' '{for(i=12;i<=NF;i++){if($i ~ /^dx:i:0$/ || $i ~ /^dx:i:-1$/){print; next}}}' | samtools view -b -o ${OUTPUT_DIR}/simplex_output.bam#!/bin/bash

# Define paths to tools and directories
BAM_DIR=/path/to/merged/bam/dir
OUTPUT_DIR=/path/to/output/dir



# Example 1: Pull out duplex reads (dx:i:1). This extracts only reads marked with the 'dx:i:1' tag indicating duplex reads.
samtools view ${BAM_DIR}/merged.bam | awk -F '\t' '{for(i=12;i<=NF;i++){if($i ~ /^dx:i:1$/){print; next}}}' | samtools view -b -o ${OUTPUT_DIR}/duplex_output.bam

# Example 2: Pull out reads for simplex reads with no duplex offsprings (dx:i:0) and simplex reads with duplex offsprings (dx:i:-1).
samtools view ${BAM_DIR}/merged.bam | awk -F '\t' '{for(i=12;i<=NF;i++){if($i ~ /^dx:i:0$/ || $i ~ /^dx:i:-1$/){print; next}}}' | samtools view -b -o ${OUTPUT_DIR}/simplex_no_offspring_output.bam

# Example 3: Specifically extract reads with dx:i:-1, indicating simplex reads with duplex offsprings.
samtools view ${BAM_DIR}/promethion_simplex_rawreads.bam | awk -F '\t' '{for(i=12;i<=NF;i++){if($i ~ /^dx:i:-1$/){print; next}}}' | samtools view -b -o ${OUTPUT_DIR}/output_test.bam

````
**Note**
At the time of performing this code, there was a limitation regarding read splitting ([see issue #327 on GitHub](https://github.com/nanoporetech/dorado/issues/327#issuecomment-1691714958)); I believe this has now been addressed in a new update.

##### Adapter removal.
Given that Dorado v0.3.1 did not have integrated adapter-removal-functionality, adapter removal for the PromethION data had to be done using a different tool.
Porechop first aligns a subset of reads (default 10000 reads) to all known ONT adapter sets; this can be memeory intensive. Adapter sets with at least one high identity match (default 90%) are deemed present in the sample and removed.

```bash
#!/bin/bash

# Define working directory
OUTPUT_DIR=/path/to/output/dir

# Define input directory
INPUT_DIR=/path/to/input/dir

# Change to the working directory
cd $WORK_DIR

# Run Porechop
porechop-runner.py -i $INPUT_DIR/simplex_rawreads.fastq.gz -o simplex_porechop_trimmed.fastq.gz

````

### Basecalling with Guppy version 6.3.4+cfaa134
#### Simplex calling

```bash
#!/bin/bash
# Code to set up GPU partition **!!!THIS MAY BE SPECIFIC TO YOUR SYSTEM!!!**
module use /home/software/tools/eb_modulefiles/all/Core
module unuse /home/software/tools/modulefiles
module load GCC/10.2.0
module load CUDA/11.1.1
module load OpenMPI/4.0.5

unset CUDA_VISIBLE_DEVICES
FAST5_DIR=/path/to/all/fast5/dir

guppy -i $FAST5_dir               \
                 -r                         \
                 -s FASTQ               \
                 -x "cuda:0"                \
                 -c dna_r10.4_e8.1_sup.cfg  \
                 --min_qscore 0             \
                 --num_callers 28             \
                 --trim_adapters
````

#### Duplex basecalling
Duplex Tools (Duplex Sequencing Tools 0.3.3 © 2021- Oxford Nanopore Technologies Ltd) can be used to identify and prepare duplex pairs for basecalling with guppy, using the pairs_from_summary command to identify pair IDs from the sequencing summary and then passed to the recommend basecall-to-basecall alignment filtering using filter_pairs.

```bash
#!/bin/bash
module load GCCcore/11.3.0
module load Python/3.10.4
INPUT_SEQSUMMARY=/path/to/raw/output.txt
OUTPUT_DIR=/path/to/output/dir
FAST5_DIR=/path/to/all/fast5/dir

#Using Duplex_tools, identify reads which are candidate pairs from the sequencing_summary file:
duplex_tools pairs_from_summary $INPUT_SEQSUMMARY $OUTPUT_DUR

#Filter through the simplex fastq data for the pairs- it will search recursively for all .fastq.gz, .fastq, and .sam/.bam file in the directory
duplex_tools filter_pairs /PATH/TO/pair_stats/pair_ids.txt /PATH/TO/final_pass.fastq.gz


#The file pair_ids_filtered.txt as prepared above can be used with the original .fast5/.pod5 files produced during a sequencing run in order to calculate high quality duplex basecalls (see duplex_guppy.sh)

````

Then, run the guppy duplex caller.

```bash
#!/bin/bash
# Code to set up GPU partition **!!!THIS MAY BE SPECIFIC TO YOUR SYSTEM!!!**
module use /home/software/tools/eb_modulefiles/all/Core
module unuse /home/software/tools/modulefiles
module load GCC/10.2.0
module load CUDA/11.1.1
module load OpenMPI/4.0.5

unset CUDA_VISIBLE_DEVICES



duplex_caller=/path/to/guppy_basecaller_duplex


$duplex_caller \
    -i $FAST5_DIR   \
    -r -s duplex_calls \
    -x 'cuda:0' -c dna_r10.4_e8.1_sup.cfg \
    --duplex_pairing_mode from_pair_list \
    --min_qscore 0   \
    --num_callers 28             \
    --trim_adapters             \
    --duplex_pairing_file /PATH/TO/pair_ids_chunk_aa.txt

````
## 2. Oxford Nanopore Reads QC
### Removing control region.
Removing the 3kb lambda DNA control sequence  ([available here](https://www.biorxiv.org/content/10.1101/2023.08.05.552089v2))

```bash
#!/bin/bash
# Load required modules
module load Java/17.0.4
module load GCCcore/11.3.0
module load Python/3.10.4

# Define directories
INPUT_DIR=/nfs/scratch/finnca/kingfish_genome/ont/minion_preassembly/1.raw/duplex
OUTPUT_DIR=/nfs/scratch/finnca/kingfish_genome/ont/minion_preassembly/2.trimmed/clean/duplex

# Change directory to the input data location
cd $INPUT_DIR

# Execute Nextflow pipeline
nextflow run hoelzer/clean \
    --input_type nano \
    --input duplex_rawreads.fastq.gz \
    --control dcs \
    -profile conda \
    --output $OUTPUT_DIR

````
### Quality control
Now we can move onto quality-control assessment.
NanoPlot (version 1.32.0,De Coster & Rademakers, 2023 ) via singularity/3.7.3 was used to assess data statistics pre- and post- filtering of the .fastq files. An older version of NanoPlot was used as the current version at time of use (1.41.6) subsamples points when plotting, and therefore the plots are not as informative, especially regarding outliers.
Once quality of your data has been assessed, you can filter it accordingly. For this, we can use porechop;

```bash
#!/bin/bash
# Activate the conda environment
# conda activate chopper

# Define directories and files
INPUT_FILE=/path/to/input/simplex_cleaned.fastq.gz
OUTPUT_FILE=/path/to/output/dir/minion_simplex_porechop_cleaned_hc50_q7.fastq.gz

# Unzip, process with chopper, and compress the output
gunzip -c $INPUT_FILE | chopper --headcrop 50 --quality 7 | gzip > $OUTPUT_FILE
````

#### Programmes and versions used:

| Program         | Version      | Citation                                                                                     |
|-----------------|--------------|---------------------------------------------------------------------------------------------|
| GCCcore         | 11.3.0       | Stallman, R. M., & GCC Developer Community. (2022). [GNU Compiler Collection 11.3.0](https://gcc.gnu.org/) |
| Python          | 3.10.4       | Python Software Foundation. (2022). [Python Language Reference](https://www.python.org/)   |
| pod5            | 0.2.4        | Oxford Nanopore Technologies. (2023). [Pod5 Format](https://github.com/nanoporetech/pod5-file-format) |
| Dorado          | 0.3.1        | Oxford Nanopore Technologies Ltd. (2022). Dorado 0.3.1.                                     |
| Duplex Tools    | 0.3.3        | Oxford Nanopore Technologies Ltd. (2021). Duplex Sequencing Tools 0.3.3.                   |
| Samtools        | 1.17         | Danecek et al. (2021). [GigaScience 10(2)](https://doi.org/10.1093/gigascience/giab008).    |
| Clean           | 0.2.0        | Lataretu et al. (2023). [bioRxiv](https://doi.org/10.1101/2023.08.05.552089).              |
| Guppy           | 6.3.4+cfaa134| Oxford Nanopore Technologies Ltd. (2022). Guppy 6.3.4.                                     |
| NanoPlot        | 1.32.0       | De Coster & Rademakers. (2023). [Bioinformatics 39(5)](https://doi.org/10.1093/bioinformatics/btad324). |
| Porechop        | 0.2.4        | Wick & Volkening. (2018). [GitHub](https://github.com/rrwick/Porechop).                    |
| chopper         | 0.6.0        | De Coster & Rademakers. (2023). [Bioinformatics 39(5)](https://doi.org/10.1093/bioinformatics/btad324). |
| Other           | ---          | Relevant tools to set up a GPU partition (specific to your system).                        |
