# Genome Assembly using MaSuRCa
## The MaSuRCA assembler
The MaSuRCA assembler combines the benefits of deBruijn graph and Overlap-Layout-Consensus assembly approaches. MaSuRCA and its toolkit contains MaSuRCA genome assembler, QuORUM error corrector for Illumina data, POLCA genome polishing software, a chromosome scaffolder (not used here), jellyfish mer counter, and MUMer aligner.
MaSuRCA is a common assembler for performing short- and long- read hybrid genome assemblies of eukaryotes, with consistently good results across studies (Tan et al., 2018,Tan et al., 2018). In brief, MaSuRCA typically works as follows: Illumina paired-end short reads are first assembled into non-ambiguous super-reads, which are then mapped to Nanopore reads to further assemble them in long, high-quality pre-mega-reads. If there are gaps between mega-reads in respect to their mapping to the Nanopore reads, these gaps are filled with the Nanopore read sequence only if the Nanopore read stretch meets some minimum criteria of coverage and quality to produce the mega-reads. If there are still gaps that cannot be merged between mega-reads due to poor quality of the Nanopore sequence, regions flanking these gaps are linked together as linking pair mates. The mega-reads and linking pairs are then assembled with Flye.
Note that for the paired end parameter, the standard deviation was set to approximately 15% of the mean following MaSuRCA recommendations (https://github.com/alekseyzimin/masurca#configuration)

```bash
# MaSuRCA config input;
PE = pe 150 22 
NANOPORE = 
EXTEND_JUMP_READS=0
GRAPH_KMER_SIZE = auto
USE_LINKING_MATES = 0
GRID_BATCH_SIZE=500000000
LHE_COVERAGE=25
LIMIT_JUMP_COVERAGE = 300
CA_PARAMETERS = cgwErrorRate=0.15 ovlRefBlockSize=40000000
CLOSE_GAPS=1
NUM_THREADS = 32
JF_SIZE = 13000000000
#With the bashscript being;
#!/bin/bash
#Using the 'simplified approach' as outlined on github;
masurca -t 32 -i $R1_ILLUMINA.fastq.gz,$R2_ILLUMINA.fastq.gz -r $NANOPORE
````

## Polishing using PILON
Assemblies, especially those generated with long-read data, usually contain residual errors.
Pilon has been the most widely used polishing tool beyond the initial release of its software (Z. Chen et al., 2021). Pilon uses read alignment analysis to identify inconsistencies between the input genome and the evidence in the reads. It then attempts to make improvements to the input genome. Multiple rounds of polishing are not recommended due to declining BUSCO scores; as such,  we can run BUSCO after every iteration to track changes. BUSCO is described below.

Pilon requires aligned reads for input:
First, we need to align our reads to our genome assembly. We can use HISAT2 to index and map the short- (Illumina) and long- (ONT) data to the assembly.
We can then use Samtools to convert .sam files to .bam files, and then sort and index the .bam files.

```bash
#!/bin/bash

# Load required modules
module purge
module load hisat2/2.1.0
module load samtools/1.10

# Define directories and file paths
WORK_DIR="path/to/your/pilon-round1"
ASSEMBLY_DIR="path/to/fasta"
SAM_OUTPUT_DIR="$WORK_DIR"

# Read data
LONG_READS="path/to/ont.fastq"
R1_ILLUMINA="path/to/illumina_r1.fastq"
R2_ILLUMINA="path/to/illumina_r2.fastq"

# Move to working directory
cd $WORK_DIR

# Indexing the assembly (commented out as it seems not intended to run each time)
# hisat2-build $ASSEMBLY_DIR assembly

# Short read mapping
hisat2 -p 12 -x assembly -1 $R1_ILLUMINA -2 $R2_ILLUMINA -S paired_alignment.sam

# Convert SAM to BAM, sort, and index
samtools view --threads 12 -b -o paired_alignment.bam paired_alignment.sam
mkdir -p Samtemp
samtools sort -o paired_alignment_sorted.bam -T Samtemp --threads 12 paired_alignment.bam
samtools index paired_alignment_sorted.bam

# Long read mapping
hisat2 -p 12 -x assembly -U $LONG_READS -S $SAM_OUTPUT_DIR/unpaired_alignment.sam

# Convert SAM to BAM, sort, and index
samtools view --threads 12 -b -o unpaired_alignment.bam $SAM_OUTPUT_DIR/unpaired_alignment.sam
mkdir -p Samtemp_l
samtools sort -o unpaired_alignment_sorted.bam -T Samtemp_l --threads 12 unpaired_alignment.bam
samtools index unpaired_alignment_sorted.bam
````
Now that we have our aligned files, we can run the first round of PILON;

```bash
#!/bin/bash

# Define variables
WORK_DIR="path/to/your/pilon-round1"
ASSEMBLY_DIR="path/to/fasta"
JAVA_MEM="200g"
JAVA_TMP_DIR="Samtemp"
PILON_JAR="path/to/pilon-1.23.jar"
PAIRED_BAM="paired_alignment_sorted.bam"
UNPAIRED_BAM="unpaired_alignment_sorted.bam"
OUTPUT_NAME="polished_genome_round_1"
THREADS=18
CHUNK_SIZE=30000

# Move to working directory
cd $WORK_DIR

# Run Pilon
java -Xmx$JAVA_MEM -Djava.io.tmpdir=$JAVA_TMP_DIR -jar $PILON_JAR \
  --genome $ASSEMBLY_DIR \
  --frags $PAIRED_BAM \
  --unpaired $UNPAIRED_BAM \
  --output $OUTPUT_NAME \
  --outdir $WORK_DIR \
  --changes --fix all --threads $THREADS --chunksize $CHUNK_SIZE
````

## Purge Haplotigs
Some parts of a genome may have a very high degree of heterozygosity. This causes contigs for both haplotypes of that part of the genome to be assembled as separate primary contigs, rather than as a contig and an associated haplotig. This is especially considered a problem for Hi-C scaffolders, such as the programme YaHS (Zhou et al., 2023), which assumes input contigs are derived from a single haplotype. The solution is to identify pairs of contigs that are syntenic and remove them.
For this, PurgeHaplotigs (version 1.1.2, Roach et al., 2018) was used. PurgeHaplotigs is a pipeline which uses mapped read coverage and blast alignments to determine which contigs to keep for the haploid assembly.

### Generate PurgeHaplotigs histogram
First, we need to generate the histogram. 

```bash
#!/bin/bash
ASSEMBLY_DIR="path/to/fasta"
UNPAIRED_BAM="unpaired_alignment_sorted.bam"
purge_haplotigs hist -b $UNPAIRED_BAM -g $ASSEMBLY_DIR -t $threads
````

We can have a look at our histogram plot.

Typically, PurgeHaplotigs would expect a bimodal histogram- with one peak for haploid level of coverage and a separate peak for diploid level of coverage- but this is one peak if the levels of duplication are very low.

```bash
#contig coverage
purge_haplotigs contigcov -i $UNPAIRED_BAM.gencov  -l 5  -m 90  -h 300  [-o '_coverage_stats.csv' ]

#Where
#-l / -low       The read depth low cutoff (use the histogram to eyeball these cutoffs)
#-h / -high      The read depth high cutoff
#-m / -mid       The low point between the haploid and diploid peaks

#purge
purge_haplotigs purge -g $ASSEMBLY_DIR 					\
				  -c coverage_stats.csv	\
					-b $UNPAIRED_BAM				\
					-o Assembly_haplotig_purged 							\
					-t 10
````

It's reccommended to check each of these steps through BUSCO; this way you can keep an eye on how the completeness of your assembly is being affected. Also, check each of these steps by assessing the general assembly stats (e.g, using https://github.com/rjchallis/assembly-stats)

