# Repeat Masking.
For this example, we will identify and mask repetitive regions in the genome. To do this, we will identify repeats by both de novo modelling and homology-based repeats.
Repeat masking is an important prior step to annotation; it is likely to overestimate gene models with an unmasked genome. To do this we can use a singularity image Dfam TE Tools Container. First, we will download the dfam container. For this dfam-tetools-1.87.sif was used-- note that the recently updated version, 1.88, uses a new version of FamDB with a new format; you can download this version here https://github.com/Dfam-consortium/TETools/releases/tag/1.7. While the singularity container comes with a Dfam library, there is a much more extentive one (~700GB!), and an old (but free) RepBase library to download:
```bash
wget "https://www.dfam.org/releases/Dfam_3.7/families/Dfam.h5.gz"
wget "https://www.dfam.org/releases/Dfam_3.7/families/Dfam.h5.gz.md5"
wget https://github.com/yjx1217/RMRB/blob/master/RepBaseRepeatMaskerEdition-20181026.tar.gz
 ````
 
This script is based on Dr. Tom Oostings script, which can be found here:  https://github.com/tomoosting/ONT_Genome_Assembly?tab=readme-ov-file#9-repeat-masking. His scripts detail how to integrate Dfam and Repbase databases into the singularity image.


Note: We used 'Actinopterygii' for this example, as the species we were working on was a fish. You can replace this with the taxonomic group you are working on.


```bash
# SETUP: Define paths and directories
export LIBDIR=/path/to/dfam/Libraries
ASSEMBLY=/path/to/assembly.fasta
REPEAT_DIR=/path/to/dfam/species_denovo_libraries
DFAM_SIF=/path/to/dfam-tetools-1.87.sif
REPEAT_OUTPUT=/output/path

# Load required module
module load singularity/3.7.3

# Step 1: Build Repeat Database using the assembly
mkdir -p $REPEAT_DIR/libraries
singularity run $DFAM_SIF BuildDatabase -name $REPEAT_DIR/libraries/repeat_db_species \
                                         -engine ncbi \
                                         $ASSEMBLY

# Step 2: Identify Transposable Elements (TEs) using RepeatModeler
singularity run $DFAM_SIF RepeatModeler -threads 40 \
                                         -LTRStruct \
                                         -database $REPEAT_DIR/libraries/

# Step 3: Extract Dfam repeats for the specific taxonomic group (Actinopterygii)
singularity exec $DFAM_SIF famdb.py -i $LIBDIR/RepeatMaskerLib.h5 \
                                      families --include-class-in-name -ad \
                                      --add-reverse-complement Actinopterygii > $REPEAT_DIR/libraries/dfam_Actinopterygii_library.fa

# Step 4: Merge Dfam and RepeatModeler libraries
cat $REPEAT_DIR/libraries/repeat_db_species.fa \
    $REPEAT_DIR/libraries/dfam_Actinopterygii_library.fa > $REPEAT_DIR/combined_library.fa

# Step 5: Mask the genome using RepeatMasker with the combined library
singularity exec $DFAM_SIF RepeatMasker -pa 40 \
                                         -dir $REPEAT_OUTPUT \
                                         -xsmall \
                                         -lib $REPEAT_DIR/combined_library.fa \
                                         $ASSEMBLY

# (Optional) Step 6: Convert soft-masked assembly to hard-masked
 sed 's/[acgt]/N/g' $REPEAT_OUTPUT/assembly.softmasked.fasta > $REPEAT_OUTPUT/assembly.hardmasked.fasta
 ````
